### PR TITLE
Fix: update engine_state when history.isolation is true (#9268)

### LINKED ISSUE #9268: 'history session' outputs 0
### Describe the bug

'history session' outputs 0, while in history there is session_id

```
❯ history session
0

❯ history -l | last 3
┏━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━┓
┃ item_id ┃       start_timestamp       ┃                command                ┃     session_id      ┃  hostname   ┃      cwd       ┃ duration ┃ exit_status ┃  idx  ┃
┣━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━━━━━━━╋━━━━━━━┫
┃   28313 ┃ 2023-05-21 15:26:40.328 UTC ┃ history session                       ┃ 1684658846675441000 ┃ Pro16.local ┃ /Users/user/cy ┃     0sec ┃           0 ┃ 26947 ┃
┃   28314 ┃ 2023-05-21 15:28:19.695 UTC ┃ $env.config.history.history_isolation ┃ 1684658846675441000 ┃ Pro16.local ┃ /Users/user/cy ┃     0sec ┃           0 ┃ 26948 ┃
┃   28315 ┃ 2023-05-21 15:29:32.094 UTC ┃ history -l | last 3                   ┃ 1684658846675441000 ┃ Pro16.local ┃ /Users/user/cy ┃     0sec ┃           0 ┃ 26949 ┃
┗━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━━━━━━━┻━━━━━━━┛
```

### How to reproduce

`history session`

### Expected behavior

I expect nu to output current session id.

### Screenshots

_No response_

### Configuration

```
version
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ version            ┃ 0.80.0                              ┃
┃ branch             ┃                                     ┃
┃ commit_hash        ┃                                     ┃
┃ build_os           ┃ macos-aarch64                       ┃
┃ build_target       ┃ aarch64-apple-darwin                ┃
┃ rust_version       ┃ rustc 1.67.1 (d5a82bbd2 2023-02-07) ┃
┃ rust_channel       ┃ stable-aarch64-apple-darwin         ┃
┃ cargo_version      ┃ cargo 1.67.1 (8ecd4f20a 2023-01-10) ┃
┃ build_time         ┃ 2023-05-17 08:56:54 +00:00          ┃
┃ build_rust_channel ┃ release                             ┃
┃ features           ┃ default, zip                        ┃
┃ installed_plugins  ┃ from parquet                        ┃
┗━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by YassineHaouzane**
# Description
Fixes #9268, the issue was due to the fact that when `history_isolation` was set to true the engine_state's `session_id` wasn't updated. 

# User-Facing Changes

None since it's a bug fix.

# Tests + Formatting

I tried to extract the block that was updating the line_editor's `session_id` to make it easier to add a test.
The test checks that after a call to the function `update_line_editor_history`, the `engine_state` and the returned `line_editor` do have the same `session_id`.

### CODE REVIEW

File: crates/nu-cli/src/repl.rs [hunk 1]
```diff
@@ -729,6 +715,44 @@ pub fn evaluate_repl(
...
+        HistoryFileFormat::Sqlite => {
+            Box::new(SqliteBackedHistory::with_file(history_path.to_path_buf()).into_diagnostic()?)
+        }
+    };
+    let line_editor = line_editor
+        .with_history_session_id(history_session_id)
+        .with_history_exclusion_prefix(Some(" ".into()))
+        .with_history(history);
+
+    store_history_id_in_engine(engine_state, &line_editor);
```
Comment (YassineHaouzane) [inline crates/nu-cli/src/repl.rs:751]: > i'm just having trouble seeing where the changes are fixing the bug 🤔

Sorry I should've mentioned that in the description !

> i'm just having trouble seeing where the changes are fixing the bug 🤔

This is the only change that I did apart from the refactor.
When we were updating the line_editor's `history_session_id` that came from `history_isolation` we did not give it the `engine_state` : 
```rust
// Setup history_isolation aka "history per session"
let history_isolation = config.history_isolation;
let history_session_id = if history_isolation {
    Reedline::create_history_session_id()
} else {
    None
};
```

Comment (amtoine) [reply crates/nu-cli/src/repl.rs:751]: > Sorry I should've mentioned that in the description !
>
no worries :wink: 

> This is the only change that I did apart from the refactor.
>
i'm confused about that...
i see this snippet of code in `crates/nu-cli/src/repl.rs` but you did not change anything about _isolation_ or _giving to the `engine_state`_ in the PR, did you? :confused:

Comment (amtoine) [reply crates/nu-cli/src/repl.rs:751]: maybe it would be more direct if you could point me directly in the diff of this PR where the real meat occurs? :yum: 

i might be just completely blind and dumb here :laughing:

Comment (YassineHaouzane) [reply crates/nu-cli/src/repl.rs:751]: > i see this snippet of code in crates/nu-cli/src/repl.rs but you did not change anything about isolation or giving to the engine_state in the PR, did you? 😕

I did not change anything about _isolation_ but I did change something about _giving to the `engine_state`_.
The main change is in this diff : 
https://github.com/nushell/nushell/pull/9616/files/bc4b375f7b9a43256a321206d88b881d1edff831#diff-665392adc0e17b5c199c59ae562f98f0673e18effdb813dc3578a1e55e191e42L130-L145

As you can see here we did not previously give the new `history_session_id` to the `engine_state`.
Which now we do in the new function that I introduced (`update_line_editor_history`) : 
https://github.com/nushell/nushell/pull/9616/files/bc4b375f7b9a43256a321206d88b881d1edff831#diff-665392adc0e17b5c199c59ae562f98f0673e18effdb813dc3578a1e55e191e42R751

Comment (amtoine) [reply crates/nu-cli/src/repl.rs:751]: ooooooooh, thank you so much :pray: 

i just completely did not notice this new line
```rust
    store_history_id_in_engine(engine_state, &line_editor);
```

Comment (amtoine) [reply crates/nu-cli/src/repl.rs:751]: @YassineHaouzane 
could you update the description of the PR for future reviewers? :innocent: 
that'd be just perfect :relieved: 

something that mentions that the PR is mainly refactoring these into functions AND adding one call to update the `engine_state` :wink:

Comment (amtoine) [reply crates/nu-cli/src/repl.rs:751]: (and you commented on your changes just on the line of interest... my bad, i was confused :scream: :laughing:)

Comment (YassineHaouzane) [reply crates/nu-cli/src/repl.rs:751]: >could you update the description of the PR for future reviewers? 😇
Done 👍

Comment (amtoine) [reply crates/nu-cli/src/repl.rs:751]: thanks a lot :blush:

**Review by amtoine — APPROVED** [review]
now that i understand the change (:wink: @YassineHaouzane) it's quite a simple change that solves the issue, so i think it's real great :yum:

**Comment by fdncred** [comment]
This fix is kind of hard to follow and i'm the one who wrote the history isolation stuff. LOL. If I understand correctly, there will always be a history_id now and it'll never be 0, is that correct?

**Comment by YassineHaouzane** [comment]
>This fix is kind of hard to follow and i'm the one who wrote the history isolation stuff. LOL.

I do agree with you, I think it's due to the refactor but I didn't find a way to add a test without doing it.

>If I understand correctly, there will always be a history_id now and it'll never be 0, is that correct?

If `history_isolation` is set to `true` the `history_id` will never be 0 and it will be set to the value that `Reedline::create_history_session_id()` returns.

**Comment by fdncred** [comment]
and if `history.isolation` is set to `false`, what does it do?

**Comment by YassineHaouzane** [comment]
It will be set to 0.

**Comment by fdncred** [comment]
ok, so it works the same as it does now?

**Comment by YassineHaouzane** [comment]
Apart from the fix yes.

**Comment by fdncred** [comment]
Gotcha. Sorry for the confusion and all the questions. Let's move forward on this. I just tested it and it seems good.
